cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TypeManager.kt

```diff
@@ -544,72 +241,199 @@ class TypeManager {
-            if (this === other) {
-                return true
-            }
-            if (other !is Ancestor) {
-                return false
+    types += superTypes.flatMap { it.getAncestors(depth + 1) }
+
+    // Since the chain starts with our type, we add ourselves to it
+    types += Type.Ancestor(this, depth)
+
+    return types
+}
+
+/** Checks, if this [Type] is either derived from or equals to [superType]. */
+fun Type.isDerivedFrom(superType: Type): Boolean {
+    // Retrieve all ancestor types of our type (more concretely of the root type)
+    val root = this.root
+    val superTypes = root.ancestors.map { it.type }
+
+    // Check, if super type (or its root) is in the list
+    return superType.root in superTypes
+}
+
+/**
+ * This computed property returns the common type in a [Collection] of [Type] objects. For example,
+ * if two types `A` and `B` both derive from the interface `C`` then `C` would be returned.
+ *
+ * More specifically, the lowest common ancestors (LCA) in a tree containing all ancestors of all
+ * types in the set is returned.
+ */
+val Collection<Type>.commonType: Type?
+    get() {
+        // If we only have one type, we can just directly return it
+        val single = this.singleOrNull()
+        if (single != null) {
+            return single
+        }
+
+        // Make sure, we only compare types of the same "kind" of type (e.g. ObjectType vs.
+        // NumericType)
+        val sameKind = this.map { it::class.simpleName }.toSet().size == 1
+        if (!sameKind) {
+            return null
+        }
+
+        // We also need to make sure that we compare the same reference depth and wrap state
+        // (which contains the pointer origins), because otherwise we need to re-create the
+        // equivalent wrap state at the end. Make sure we only have one wrap state before we
+        // proceed.
+        val wrapStates = this.map { it.wrapState }.toSet()
+        val wrapState = wrapStates.singleOrNull() ?: return null
+
+        // Build all ancestors out of the root types. This way we compare the most inner type,
+        // regardless of the wrap state.
+        val allAncestors = this.map { it.root.ancestors }
+
+        // Because the ancestors all have different "depths", we need to normalize them before
+        // continuing. In this case we are reversing the order of the depths, setting the root node
+        // with the largest depth chain to 0 and then increase the depth with each level.
+        for (ancestors in allAncestors) {
+            // Calculate the maximum depth
+            val maxDepth = ancestors.maxBy(Type.Ancestor::depth).depth
+            // Subtract the current depth from the maximum depth to reverse the order and start at 0
+            // for the root node that is the farthest away
+            ancestors.forEach { it.depth = (maxDepth - it.depth) }
+        }
+
+        // Find the lowest common ancestor (LCA) by:
```

Comment:
Note to self: this is not 100 % accurate all the time and needs to be changed back to a proper LCA algorithm